src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+
+            max_semantic_history = int(np.floor(max_coarse_history / semantic_to_coarse_ratio))
+            # trim histories correctly
+            n_semantic_hist_provided = np.min(
+                [
+                    max_semantic_history,
+                    x_semantic_history.shape[1] - x_semantic_history.shape[1] % 2,
+                    int(np.floor(x_coarse_history.shape[1] / semantic_to_coarse_ratio)),
+                ]
+            )
+
+            n_coarse_hist_provided = int(round(n_semantic_hist_provided * semantic_to_coarse_ratio))
+
+            x_semantic_history = x_semantic_history[:,-n_semantic_hist_provided:].astype(np.int32)
+            x_coarse_history = x_coarse_history[:, -n_coarse_hist_provided:].astype(np.int32)
+            # bit of a hack for time alignment (sounds better) - from Bark original implementation
+            x_coarse_history = x_coarse_history[:,:-2]
+
+        else:
+            # shape: (batch_size, 0)
+            x_semantic_history = np.array([[]]*batch_size, dtype=np.int32)
+            x_coarse_history = np.array([[]]*batch_size, dtype=np.int32)
+
+        x_semantic_history = torch.from_numpy(x_semantic_history).to(self.device)
+        x_coarse_history = torch.from_numpy(x_coarse_history).to(self.device)
+
+        return x_semantic_history, x_coarse_history
+
+    def generate_text_semantic(
+        self,
+        inputs: torch.Tensor,
+        history_prompt: Optional[Dict[str, np.ndarray]] = None,
+        **kwargs,
+    ) -> torch.LongTensor:
+        # TODO: add a max_gen_duration_s early stop
+
+        input_ids = inputs
+
+        # TODO: Not used for now. where to set the default value ?
+        # min_eos_p = kwargs.get("min_eos_p", 0.2)
+
+        # TODO: input_ids[:,256:256+256] (to verify) corresponds to history_prompt["semantic_prompt"], maybe use that
+        # input_ids should be of shape (batch_size, seq_len) where seq_len = 513
+        input_embeds = torch.cat(
+            [
+                self.semantic_model.transformer.wte(input_ids[:, :256])
+                + self.semantic_model.transformer.wte(input_ids[:, 256 : 256 + 256]),
+                self.semantic_model.transformer.wte(input_ids[:, 256 + 256 :]),
+            ],
+            dim=1,
+        )
+
+        semantic_logits_processor = SemanticLogitsProcessor(
+            self.config.semantic_vocab_size, self.config.semantic_pad_token
+        )
+
+        # TODO: for now, it is not implemented yet as long as StoppingCriteria issue is not dealt with
+        # https://github.com/huggingface/transformers/issues/23674
+        # semantic_stopping_criteria = SemanticStoppingCriteria(min_eos_p , self.config.semantic_pad_token)
+
+        # pass input_ids in order to stay consistent with the transformers generate method even though it is not used (except to get the input seq_len - that's why we keep the first 257 tokens)
+        semantic_output = self.semantic_model.generate(
+            input_ids[:, :257],
```

Comment:
This hard coding can also then be replaced

Reply:
how  come ?